File: Applied Course/4.Problem Solving/11.Problem on Dynamic Programming/7.Coin Change Problem.py
# Logic-1 recursion Solution
import sys
def minCoinChange(arr,amt,length):
    if amt == 0:
        return 0
    res = sys.maxsize
    for i in range(length):
        if arr[i]<=amt:
            temp=minCoinChange(arr,amt-arr[i],length)
            
            if temp+1<res:
                res=temp+1
    return res

if __name__=='__main__':
    arr = [1, 2, 3] 
    m = len(arr) 
    print(minCoinChange(arr, 15, m))


# Logic-2 bottom up approach
import sys
def minCoinChangeDP(arr,amt,m):
    temp=0
    dp=[0 for i in range(amt+1)]
    for i in range(1,amt+1):
        dp[i]=sys.maxsize
    for i in range(1,amt+1):
        for j in range(m):
            if arr[j]<=i:
                temp=dp[i-arr[j]]
            if temp+1<dp[i]:
                dp[i]=temp+1
    return dp[amt]
if __name__=='__main__':
    arr = [1, 2, 3] 
    m = len(arr) 
    print(minCoinChangeDP(arr, 15, m))

File: Applied Course/4.Problem Solving/11.Problem on Dynamic Programming/7.Coin Change Problem.cpp
#include<bits/stdc++.h>
using namespace std;
int solve(int arr[],int n,int amt)
{
	int temp=0;
	int dp[101]={0};
	for(int i=1;i<amt+1;i++)
		dp[i]=INT_MAX;
	for(int i=1;i<amt+1;i++)
	{
		for(int j=1;j<n;j++)
		{
			if(arr[j]<=i)
                temp=dp[i-arr[j]];
            if(temp+1<dp[i])
                dp[i]=temp+1;
		}
	}
	return dp[amt];
}
int main()
{
	int arr[]={1,2,3};
	int n=sizeof(arr)/sizeof(arr[0]);
	cout<<solve(arr,n,15);
	return 0;
}